### ISSUE TITLE
Presumptuous monadism

### ISSUE THREAD

**Issue by stuartpb** [bmillwood/pointfree#24]
I passed this input to pointfree.io:

```haskell
latterLetter1d alphabet letter = (elemIndex letter alphabet) `further` latter alphabet
```

This was the output:

```haskell
latterLetter1d = ap (flip . (further .) . flip elemIndex) latter
```

Is there no pointfree way to express this *without* importing Control.Monad? Or, if monads are really the cleanest way to do this, is there a formulation of it without so many dots and flips (#23)?

**Comment by josephcsible** [comment]
> Is there no pointfree way to express this without importing Control.Monad?

Not to my knowledge. I don't know of any duplicative combinators in the Prelude. (Nitpick: technically you could use liftA2 instead of liftM2, but then you just have to import Control.Applicative instead of Control.Monad.)

> Or, if monads are really the cleanest way to do this, is there a formulation of it without so many dots and flips (#23)?

Yes:

```haskell
latterLetter1d = flip (flip (liftM2 further) latter . elemIndex)
```

**Comment by bmillwood** [comment]
Nowadays I think you don't need to import Control.Monad to get these instances, so I don't think it's bad to use the monad interface here.

**Comment by josephcsible** [comment]
No, you still do need to import Control.Monad, even in GHC 8.8.1.

**Comment by josephcsible** [comment]
Hello?

**Comment by bmillwood** [comment]
oh, I misunderstood this at the time, sorry (just happened to revisit it now). The issue is not the instance itself, but the function `ap`.

Nevertheless, I don't personally feel excited to work on this. It seems like using standard base imports is often going to result in better code, so at most I'd want this to be an option, `--prelude-only` or something.

**Comment by josephcsible** [comment]
> It seems like using standard base imports is often going to result in better code

For other things like `join` and `liftA2` I agree, but for `ap` in particular, is there ever a reason to use it rather than `<*>`?
